design-docs/2021/q1/credrank-migration.md

```diff
@@ -0,0 +1,97 @@
+# CredRank Migration Design Doc
+
+## Last Updated: 2021.01.07
+
+## Authors:
+ - @decentralion
+
+
+## Overview
+
+We've finished implementing [CredRank], but we haven't yet migrated SourceCred to use it.
+This doc outlines a plan for migrating all of SourceCred to use Cred scores from CredRank,
+and to remove [Timeline Cred].
+
+[CredRank]: https://discourse.sourcecred.io/t/credrank-scalable-interpretable-flexible-attribution/654
+[Timeline Cred]: https://github.com/sourcecred/sourcecred/pull/1212
+
+## Steps
+
+### Finish Implementing CredGrainView
+
+The `CredGrainView` (see src/core/credGrainView.js) is a class that makes it
+convenient to retrieve Cred (and Grain) data about any participant. It will act
+as the main data API layer for consumers that need access to Cred data. It's
+thus a replacement for the existing `CredView` class at
+src/analysis/credView.js.
+
+Currently, CGV exists as an interface, but most of the methods are not
+implemented and just throw errors. It needs to be implemented and tested.
+
+### Create a CredRank-compatible Cred Explorer
+
+Presently, the Explorer (src/ui/components/Explorer/Explorer.js) depends on the
+`CredView` class, which means it uses TimelineCred scores. We should create a
+new overview page that depends on `CredGrainView` instead.
+
+Less ambitiously, we could just port the existing UI over to use `CredView`.
+More ambitiously, we could implement the new frontend described in [#2227]. I
+think it's probably better to port the existing UI and then treat upgrading to
+the new frontend as a separate project, so that we can land this migration more
+quickly and easily.
+
+[#2227]: https://github.com/sourcecred/sourcecred/issues/2227
+
+The new Explorer should be wired in as a hidden route, e.g.
+`/#/explorer-credrank`. While it's in the hidden state, it will still be
+available for debugging/testing purposes, but only for those who know the route
+exists, and manually ran the cli `credrank` command first.
+
+### Create a CredRank-compatible `grain` command
+
+Presently, the `grain` command in src/cli/grain.js depends on
+`core/ledger/applyDistributions`, which itself depends on `CredView`. Thus,
+Grain distributions currently operate on TimelineCred. We should write a new
+version of the Grain command (which we can call `grain2` or `grain-credrank`)
+which uses the `CredGrainView` instead.
+
+Once the new Grain command is available, it should be wired in to the CLI for
+testing, but not actively used until the migration.
+
+### Migrate
+
+Once we have a replacement UI and Grain command that use the `CredGrainView`,
+we will be ready to atomically migrate over to the new algorithm. This
+migration will have three steps:
+
+1. In src/cli/go.js, replace the `score` command with the `credrank` command in
```

Comment:
nit, I would probably choose to rename `credrank` to `score` and rename `score` to `timeline-cred` (or remove the timeline-cred CLI) to avoid UX churn.

Reply:
I think it might be even better to just expose `credrank` and `timeline-cred` (or `timelineCred`?) in the lower-level CLI explicitly.

This wouldn't necessarily induce much churn for users at least, since these commands aren't exposed by name in the instance scripts. It might be a bit of change for developers but we could alias `score` to `credrank` if that's indeed a concern.

Reply:
I like credrank better than score. Mostly because I think the CredRank algorithm is really cool, so I'd rather surface that cool name than the more generic and boring "score".

As @topocount pointed out, this won't induce much churn for users because we're telling users to use `sourcecred go` rather than run lower-level commands directly; the lower-level commands become more of a core dev affordance.
 However, there are some users out there that still have pointers to `score`, e.g. [1hive/pollen](https://github.com/1Hive/pollen). So we should either help them upgrade or leave score as an alias.

Reply:
Oh, and there's no reason to keep `timeline-cred` as a command; we're killing timeline cred with fire 🔥